addons/sale_loyalty/models/sale_order.py

```diff
@@ -631,11 +631,13 @@ def _update_programs_and_rewards(self):
 
         # Automatically load in eWallet coupons
         if self._allow_nominative_programs():
-            ewallet_coupons = self.env['loyalty.card'].search(
-                [('id', 'not in', self.applied_coupon_ids.ids), ('partner_id', '=', self.partner_id.id),
-                ('points', '>', 0), ('program_id.program_type', '=', 'ewallet')])
-            if ewallet_coupons:
-                self.applied_coupon_ids += ewallet_coupons
+            coupons = self.env['loyalty.card'].search([
+                ('partner_id.commercial_partner_id', '=', self.partner_id.commercial_partner_id.id),
+                ('points', '>', 0),
+                ('program_id.program_type', 'in', ('ewallet', 'loyalty'))
+            ])
+            if coupons:
+                self.applied_coupon_ids |= coupons
```

Comment:
Didn't we agree on this? :smiling_imp:
(directly assign + missing comma) 
```suggestion
            self.applied_coupon_ids |= self.env['loyalty.card'].search([
                ('partner_id.commercial_partner_id', '=', self.partner_id.commercial_partner_id.id),
                ('points', '>', 0),
                ('program_id.program_type', 'in', ('ewallet', 'loyalty')),
            ])
```